webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala

```diff
@@ -262,32 +287,91 @@ trait CachedFindRuleNodeStatusReports extends ReportingService with CachedReposi
+      // * we have a compliance: yeah, put it in the cache
+      // * new policy generation, a new nodeexpectedreports is available: compute compliance for last run of the node, based on this nodeexpectedreports
+      // * node deletion: remove from the cache
+      // * invalidate cache: ???
+      // * no report from the node (compliance expires): recompute compliance
+
+      // as a first approach, they could simply findRuleNodeStatusReports
+      {
+        (for {
+          _  <- performAction(actions)
+        } yield ()).catchAll(err => ReportLoggerPure.Cache.error(s"Error when updating compliance cache for nodes: [${actions.map(_.nodeId).map(_.value).mkString(", ")}]: ${err.fullMsg}"))
+      }
     )
   )
 
   // start updating
   updateCacheFromRequest.forever.forkDaemon.runNow
 
 
+  /**
+   * Do something with the action we received
+   * All actions must have *exactly* the *same* type
+   */
+  private[this] def performAction(actions: List[CacheComplianceQueueAction]): IOResult[Unit] = {
+    ReportLoggerPure.Cache.debug(s"Performing action ${actions.headOption}") *>
+    // get type of action
+      (actions.headOption match {
+      case None => ReportLoggerPure.Cache.debug("Nothing to do")
+      case Some(t) => t match {
+        case update:UpdateCompliance =>
+          ReportLoggerPure.Cache.debug(s"Compliance cache updated for nodes: ${actions.map(_.nodeId.value).mkString(", ")}") *>
+          IOResult.effectNonBlocking {
+            cache = cache ++ actions.map { case x: UpdateCompliance => (x.nodeId, x.nodeCompliance) }
+          }
+        case delete: RemoveNodeInCache =>
+          IOResult.effectNonBlocking {
+            cache = cache.removedAll(actions.map { case x: RemoveNodeInCache => (x.nodeId) })
+          }
+
+        // need to compute compliance
+        case _ =>
+          val impactedNodeIds = actions.map(x => x.nodeId)
+          for {
+              x <- ZIO.foreach(impactedNodeIds.grouped(batchSize).to(Seq)) { updatedNodes =>
+                  for {
+                  updated <- defaultFindRuleNodeStatusReports.findRuleNodeStatusReports(updatedNodes.toSet, Set()).toIO
+                  _       <- IOResult.effectNonBlocking {
+                  cache = cache ++ updated
+                }
+                _ <- ReportLoggerPure.Cache.debug(s"Compliance cache recomputed for nodes: ${updated.keys.map(_.value).mkString(", ")}")
+                } yield ()
+            }
+          } yield ()
+
+
+          /*for {
+            updated <- defaultFindRuleNodeStatusReports.findRuleNodeStatusReports(impactedNodeIds.toSet, Set()).toIO
+            _       <- IOResult.effectNonBlocking {
+              cache = cache ++ updated
+            }
+            _ <- ReportLoggerPure.Cache.debug(s"Compliance cache recomputed for nodes: ${updated.keys.map(_.value).mkString(", ")}")
+          } yield ()*/
+      }
+    })
+  }
+  /**
+   * Group all actions queue by the same type, keeping the global order.
+   */
```

Comment:
I don't understand how order is kept. IIUC, if you have that in queue:
````
List(
  InsertNodeInCache(node1)
, RemoveNodeInCache(node1) 
, InsertNodeInCache(node1)
)
```
It will becomes: 
```
List(
  List(
    InsertNodeInCache(node1)
  , InsertNodeInCache(node1)
  )
, List(
    RemoveNodeInCache(node1) 
  )
)
``` 
Which is semantically different.

Reply:
no, it does not reorder, it create list of list in same order
`
List(
  InsertNodeInCache(node1)
, InsertNodeInCache(node2)
, RemoveNodeInCache(node1)
, InsertNodeInCache(node2)
, InsertNodeInCache(node1)
)

becomes

List(
  List(
    InsertNodeInCache(node1)
  , InsertNodeInCache(node2)
  )
, List(
    RemoveNodeInCache(node1) 
  )
 , List(
    InsertNodeInCache(node2)
  , InsertNodeInCache(node1)
  )
`

Reply:
AH OK :) I could be smarter in a second time, and group same event as long as they belong to different nodes, and simplify action (like insert / delete / insert => insert). Lots of opportunities now that we have a sound ground :)

Reply:
we don't want to be clever on that, because we don't want to loose data when stored in compliancelevels 
(or the necessary cleverness would be too hard to reach a correct result)

Reply:
we will, at some point (for perf or whatever). Still, the comment about keeping all intermediate compliance for historization is very interesting and it must be put here!